File: ThreemaFramework/Business/TypingIndicatorManager.m
#import "TypingIndicatorManager.h"
#import "ThreemaFramework/ThreemaFramework-Swift.h"

#ifdef DEBUG
  static const DDLogLevel ddLogLevel = DDLogLevelVerbose;
#else
  static const DDLogLevel ddLogLevel = DDLogLevelWarning;
#endif

@implementation TypingIndicatorManager {
    dispatch_source_t resetTimer;
    dispatch_queue_t resetQueue;
    BOOL timerSuspended;
}

+ (int) typingIndicatorResendInterval {
    return kTypingIndicatorResendInterval;
}

+ (int) typingIndicatorTypingPauseInterval {
    return kTypingIndicatorTypingPauseInterval;
}

+ (TypingIndicatorManager*)sharedInstance {
    static TypingIndicatorManager *sharedInstance;
    static dispatch_once_t pred;
    dispatch_once(&pred, ^{
        sharedInstance = [[TypingIndicatorManager alloc] init];
    });
    return sharedInstance;
}

- (id)init
{
    self = [super init];
    if (self) {
        resetQueue = dispatch_queue_create("ch.threema.resetQueue", 0);
        resetTimer = dispatch_source_create(DISPATCH_SOURCE_TYPE_TIMER, 0, 0, resetQueue);
        dispatch_source_set_timer(resetTimer, dispatch_time(DISPATCH_TIME_NOW, kTypingIndicatorTimeout/2 * NSEC_PER_SEC),
                                  kTypingIndicatorTimeout/2 * NSEC_PER_SEC, NSEC_PER_SEC);
        dispatch_source_set_event_handler(resetTimer, ^{
            [self resetTypingIndicators];
        });
        dispatch_resume(resetTimer);
    }
    return self;
}

- (void)startObserving {
    if (!timerSuspended) {
        return;
    }
    
    dispatch_resume(resetTimer);
    timerSuspended = NO;
    DDLogVerbose(@"Typing indicator observing started");
}

- (void)stopObserving {
    if (timerSuspended) {
        return;
    }
    
    dispatch_suspend(resetTimer);
    timerSuspended = YES;
    DDLogVerbose(@"Typing indicator observing stopped");
}

- (void)resetTypingIndicators {
    DDLogVerbose(@"Resetting typing indicators");
    dispatch_async(dispatch_get_main_queue(), ^{
        /* Fetch all Conversations that are currently typing, and reset the typing
         indicator if it was received too long ago */
        EntityManager *entityManager = [[BusinessInjector ui] entityManager];
        [entityManager performAndWaitSave:^{
            NSDate *timeoutDate = [NSDate dateWithTimeIntervalSinceNow:-kTypingIndicatorTimeout];
            NSArray *conversations = [entityManager.entityFetcher typingConversationEntitiesWithTimeoutDate:timeoutDate];
            
            if (conversations == nil) {
                return;
            }

            for (ConversationEntity *conversation in conversations) {
                DDLogVerbose(@"Reset typing indicator on conversation with %@", conversation.contact.identity);
                conversation.typing = @NO;
            }
        }];
    });
}

- (void)setTypingIndicatorForIdentity:(NSString*)identity typing:(BOOL)typing {
    DDLogInfo(@"Started setting typing indicator `%@` for conversation with contact identity %@.", typing ? @"ON" : @"OFF", identity);

    if (identity.length != ThreemaIdentityObjc.length) {
        DDLogError(@"Invalid contact identity: %@.", identity);
        return;
    }

    dispatch_async(dispatch_get_main_queue(), ^{
        EntityFetcher *entityFetcher = [[BusinessInjector ui] entityManager].entityFetcher;

        ConversationEntity *conversation = [entityFetcher conversationEntityFor:identity];
        if (conversation == nil) {
            DDLogError(@"No conversation found with contact identity %@.", identity);
            return;
        }

        conversation.typing = [NSNumber numberWithBool:typing];
        DDLogInfo(@"Typing indicator for conversation with contact identity %@ successfully set.", identity);
    });
}

@end

File: ThreemaFramework/Business/TypingIndicator/TypingIndicatorManager.swift
import CocoaLumberjackSwift
import Foundation
import ThreemaEssentials

@objcMembers public final class TypingIndicatorManager: NSObject {
    // MARK: - Public properties

    public static let sharedInstance = TypingIndicatorManager()

    // MARK: - Private properties

    private let resetQueue = DispatchQueue(label: TypingIndicatorConfig.resetQueueLabel)
    private var resetTimer: DispatchSourceTimer?
    private var timerSuspended = false

    // MARK: - Lifecycle

    override public init() {
        super.init()
        setupResetTimer()
    }

    deinit {
        resetTimer?.cancel()
    }

    // MARK: - Public methods

    public func startObserving() {
        guard timerSuspended else {
            return
        }
        
        resetTypingIndicators()
        
        resetTimer?.resume()
        timerSuspended = false
        DDLogVerbose("Typing indicator observing started")
    }

    public func stopObserving() {
        guard !timerSuspended else {
            return
        }
        resetTimer?.suspend()
        timerSuspended = true
        DDLogVerbose("Typing indicator observing stopped")
    }

    public func resetTypingIndicators() {
        DispatchQueue.main.async {
            DDLogInfo("Started resetting the typing indicators for all conversations.")

            // Fetch all Conversations that are currently typing, and reset the typing
            // indicator if it was received too long ago
            let timeoutDate = Date(timeIntervalSinceNow: -TypingIndicatorConfig.timeoutInterval)

            let entityManager = BusinessInjector.ui.entityManager
            let entityFetcher = entityManager.entityFetcher

            entityManager.performAndWaitSave {
                let conversations = entityFetcher.typingConversationEntities(timeoutDate: timeoutDate)
                guard !conversations.isEmpty else {
                    DDLogInfo("No conversations eligible for resetting the typing indicator.")
                    return
                }
                conversations.forEach { $0.typing = NSNumber(value: false) }
                DDLogInfo("Typing indicator has been successfully reset for \(conversations.count) conversation(s).")
            }
        }
    }

    public func setTypingIndicator(for identity: String, typing: Bool) {
        DDLogInfo(
            "Started setting typing indicator `\(typing ? "ON" : "OFF")` for conversation with contact identity \(identity)."
        )

        guard identity.count == ThreemaIdentity.length else {
            DDLogError("Invalid contact identity: \(identity).")
            return
        }

        DispatchQueue.main.async {
            let entityFetcher = BusinessInjector.ui.entityManager.entityFetcher

            guard let conversation = entityFetcher.conversationEntity(for: identity) else {
                DDLogError("No conversation found with contact identity \(identity).")
                return
            }

            conversation.typing = NSNumber(value: typing)
            DDLogInfo("Typing indicator for conversation with contact identity \(identity) successfully set.")
        }
    }

    // MARK: - Private methods

    private func setupResetTimer() {

        // Prevent duplicate timers

        guard resetTimer == nil else {
            return
        }

        resetTimer = DispatchSource.makeTimerSource(queue: resetQueue)

        resetTimer?.schedule(
            wallDeadline: .now() + TypingIndicatorConfig.resetCheckInterval,
            repeating: TypingIndicatorConfig.resetCheckInterval
        )

        resetTimer?.setEventHandler { [weak self] in
            self?.resetTypingIndicators()
        }
        resetTimer?.resume()
    }
}
